googlemock/test/gmock-function-mocker_test.cc

```diff
@@ -656,5 +657,27 @@ TEST(MockMethodMockFunctionTest, MockMethodSizeOverhead) {
   EXPECT_EQ(sizeof(MockMethodSizes0), sizeof(MockMethodSizes4));
 }
 
+void hasTwoParams(int, int);
+struct MockMethodNoexceptSpecifier {
+  MOCK_METHOD(void, func1, (), (noexcept));
+  MOCK_METHOD(void, func2, (), (noexcept(true)));
+  MOCK_METHOD(void, func3, (), (noexcept(false)));
+  MOCK_METHOD(void, func4, (), (noexcept(noexcept(1+1))));
+  MOCK_METHOD(void, func5, (), (const, noexcept(noexcept(1+1))));
+  MOCK_METHOD(void, func6, (), (noexcept(noexcept(1+1)), const));
+  // Put commas in the noexcept expression
+  MOCK_METHOD(void, func7, (), (noexcept(noexcept(hasTwoParams(1,2))), const));
+};
+
+TEST(MockMethodMockFunctionTest, NoexceptSpecifierPreserved) {
+  EXPECT_TRUE(noexcept(std::declval<MockMethodNoexceptSpecifier>().func1()));
```

Comment:
Is there a reason to use declval here? I'd simply define a MockMethodNoexceptSpecifier object.
In case the compiler warns that it's unused, it can be suppressed by cast to void.

Reply:
I had that previously, but the compiler would always say the result of the expression is `noexcept(false)`.
https://travis-ci.org/google/googletest/jobs/594800623#L2313
My guess would be it has to do with the constructor not being marked `noexcept`.
But `declval` really is the "correct" solution as we just want to test the `noexcept` specifier for the function (and not the default constructor as well).

Another solution would be to use `std::is_nothrow_invocable`, but I don't think there's a technical reason to prefer one over another.

Reply:
My intention was to define one MockMethodNoexceptSpecifier object in the test, and use it in all expectations. Then it wouldn't matter if the constructor is noexcept.

Reply:
That's definitely another solution. Personally, I wouldn't like having to declare an instance of an object (which is a runtime construct) just so syntactically (since `noexcept` expressions are evaluated at compile-time) I can reason about the methods. 
But I also can't argue it would be less code.
If this comes up in the "internal review" I'd be alright changing it.

Reply:
Yeah, it's all about tradeoffs if you want to use `static_assert`s in your test code to test some logic. 
If you move it into a compile-time check, you know it'll always be true. The tests wouldn't compile otherwise. On the other hand, normally one test failing doesn't bar other tests from running and reporting their successes/failures. Since the `noexcept` spec wouldn't be absolutely fatal if it was wrong, I opted to implement them as runtime checks so they can report successes/failures just like other tests.